### DISCUSSION TITLE
H1 and L2 norms on Hermite space

### DISCUSSION THREAD

**Discussion by nadav7679** [firedrakeproject/firedrake#3747] [Firedrake support]
Hi,

I'm trying to calculate the H1 norm of functions on Hermite spaces (1D intervals) using the coefficients of the functions, i.e. using `dat.data`. The way I understand it, the coefficients in `dat.data` of a function in Hermite space consist of point-evaluations and derivative point-evaluations at edges of equidistant cells. Therefore, they can be used in Riemann sums to approximate both the L2 norm and the L2 norm of the derivative, so that we'd get the H1 norm.

However, this does not seem to be the case when implementing:

```python3
N = 2048
L = 10
h = L/N

mesh = PeriodicIntervalMesh(ncells=N, length=L)
x = SpatialCoordinate(mesh)[0]
her = FunctionSpace(mesh, "HER", 3) 

sin_func = Function(her)
sin_func.project(sin(2 * pi *x /L))

fd_L2_norm = norm(sin_func, "L2")
fd_H1_norm = norm(sin_func, "H1")
coeff_norm = np.sqrt(h * np.sum(sin_func.dat.data ** 2))

print(f"Firedrake L2: {fd_L2_norm}")
print(f"Firedrake H1: {fd_H1_norm}")
print(f"Riemann sum (H1): {coeff_norm}")
```  
Output:

```
Firedrake L2: 2.236067977499789
Firedrake H1: 2.640818221729369
Riemann sum (H1): 2.23607851112194
```

Both the Riemann sum and the Firedrake L2 give a correct result for the L2 norm, i.e. $\sqrt{5}$, and the H1 norm is correct as well (according to Wolfram). 

I would have expected the Riemann sum to agree with the H1 norm, not the L2 one. I'm obviously missing something, so I'd be grateful if anyone can offer an explanation.

Thanks!

**Comment by dham** [comment]
The derivative point evaluations are expressed in reference coordinates and need to be pulled back to physical space before you can use them in the way you are suggesting. They will be about 200 (L/N) times too small in the calculation. Notice that the L2 norm and your sum differ in the 6th significant figure. This is too large to be roundoff but is about consistent with adding the reference gradient.

**Reply by nadav7679** [reply]
Thank you, I see that now.

I tried implementing your suggestion and it works. I computed this lovely expression:
```python3

h ** 0.5 * np.sqrt(np.sum(sin_func.dat.data[0::4] ** 2) + \
                   np.sum((pullback_ratio * sin_func.dat.data[1::4]) ** 2) + \
                   np.sum(sin_func.dat.data[2::4] ** 2) + \
                   np.sum((pullback_ratio*sin_func.dat.data[3::4]) ** 2)
                  )
```
which produces the correct answer with 

```python3
pullback_ratio = abs((np.pi/5)/min(sin_func.dat.data[1::4]))
```
as the minimum in this example should be exactly $-\frac{\pi}{5}$.

Now that I look at it, I believe the `pullback_ratio` is simply `h`, so I should be able to work with it :)
